### PR TITLE
All my things

### PR DESCRIPTION
**Pull request by djspiewak**
Merge this one, not the others.  :-D

### CODE REVIEW

File: connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala [hunk 1]
```diff
@@ -16,103 +16,272 @@
...
+  private def updateGraph[
+      G[_]: Monad: NameGenerator: MonadState_[?[_], RevIdx]: MonadState_[?[_], QSUDims[T]]: PlannerErrorME](
+      pat: QScriptUniform[Symbol]): G[QSUGraph] = {
+
+    for {
+      qgraph <- QSUGraph.withName[T, G](pat)
 
-    minimized
+      dims <- MonadState_[G, QSUDims[T]].get
+      computed <- AP.computeProvenanceƒ[G].apply(QSUGraph.QSUPattern(qgraph.root, pat.map(dims)))
```
Comment (wemrysi) [inline connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala]: This is necessary, but I don't think it is sufficient. Other vertices downstream from one of the merged reductions will have dimensions referencing the bucket(s) of that reducer. Those dimensions either need to be remapped to refer to the buckets of the merged reducer or just recomputed, which will have the same effect.

File: connector/src/main/scala/quasar/qscript/qsu/LPtoQS.scala [hunk 2]
```diff
@@ -63,15 +63,21 @@ final class LPtoQS[T[_[_]]: BirecursiveT: EqualT: ShowT] extends QSUTTypes[T] {
...
 
   private def debugAG[F[_]: Applicative](prefix: String): K[F, AuthenticatedQSU[T], AuthenticatedQSU[T]] =
     K { aqsu =>
-      println("\n\n" + prefix + aqsu.graph.shows + "\n" + aqsu.dims.shows)
+      maybePrint("\n\n" + prefix + aqsu.graph.shows + "\n" + aqsu.dims.shows)
       aqsu.point[F]
     }
+
+  private def maybePrint(str: String): Unit = {
+    // println(str)
```
Comment (wemrysi) [inline connector/src/main/scala/quasar/qscript/qsu/LPtoQS.scala]: Thanks for cleaning up the mess I left :cold_sweat:

File: connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala [hunk 3]
```diff
@@ -16,103 +16,272 @@
...
-      case minimized =>
-        // TODO implement coalesce heuristics here
+    case candidates =>
+      val reducerAttempt = candidates collect {
+        case g @ QSReduce(source, buckets, reducers, repair) =>
+          (source, buckets, reducers, repair)
+      }
+
+      // candidates.forall(_ ~= QSReduce)
+      if (reducerAttempt.lengthCompare(candidates.length) === 0) {
```
[the remark this answers is not in the document]
Comment (djspiewak) [reply connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala]: I thought about it, but we actually have the guarantee not only that the list is non-empty, but it's actually length >= 2.

File: connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala (hunk 1 above)
Comment (djspiewak) [reply connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala]: The buckets of the merged reducer will be equivalent to the buckets of the unmerged reducer, by definition.

Comment (wemrysi) [reply connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala]: True, but for two reducer nodes, `'a` and `'b`, prior to the merge each would have dimensions like `[Value(Bucket('a, 0)), Proj(foo)]` and `[Value(Bucket('b, 0)), Proj(foo)]` so if `'b` was merged into `'a` then it's bucket references by dimensions of other vertices need to be remapped to `'a`.

`QProv.rename` should do the trick here.

Comment (djspiewak) [reply connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala]: Ah interesting. I’ll fix that

**Comment by jsantos17** [comment]
Is there anything missing to merge this? It fixes a bunch of Mongo planner issues and I need it to start fixing the tests for Mongo in `qsu`.

**Pull request merged**